test/jdk/java/foreign/TestDeadlock.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @run main/othervm/timeout=5 --enable-native-access=ALL-UNNAMED -Xlog:class+init TestDeadlock
+ */
+
+import java.lang.foreign.*;
+import java.util.concurrent.CountDownLatch;
+
+public class TestDeadlock {
+    public static void main(String[] args) throws Throwable {
+        CountDownLatch latch = new CountDownLatch(2);
+
+        Thread t1 = Thread.ofPlatform().start(() -> {
+            Arena arena = Arena.global();
+            arena.scope(); // init ArenaImpl
+            ValueLayout.JAVA_INT.byteSize(); // init ValueLayout (and impls)
+            latch.countDown();
+            try {
+                latch.await();
+            } catch(InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+
+            // Access ArenaImpl -> NativeMemorySegmentImpl -> MemorySegment
+            arena.allocateFrom(ValueLayout.JAVA_INT, 42);
+        });
+
+        Thread t2 = Thread.ofPlatform().start(() -> {
```

Comment:
We do not seem to test:
* mapped segments
* symbol lookup segments
* linker-generated segments

Is that ok?

Reply:
I'll try to add a test for mapped segments as well (through FileChannel::map). Though, it might not be possible to reproduce an issue in a test. 

But, the other two do not touch NativeMemorySegmentImpl or MappedMemorySegmentImpl directly, so I'm not sure. Theoretically we can test all the API points (that create MemorySegments), but given the tricky timing (the current test has to carefully pre-initialize several classes to get there), I'm not sure it's useful to spend resources on running those tests.

Reply:
I added a similar test for FileChannel::map, but was not able to reproduce a deadlock. The log shows that the basic conditions are right, but I guess we just miss the timing:

```
[0.182s][info][class,init] 728 Initializing 'java/lang/foreign/MemorySegment' (0x0000021a82050320) by thread "Thread-2"
[0.183s][info][class,init] 729 Initializing 'jdk/internal/foreign/AbstractMemorySegmentImpl' (0x0000021a82050d28) by thread "Thread-2"
[0.183s][info][class,init] 730 Initializing 'jdk/internal/foreign/NativeMemorySegmentImpl' (0x0000021a82051598) by thread "Thread-2"
[0.183s][info][class,init] 731 Initializing 'jdk/internal/foreign/MappedMemorySegmentImpl' (0x0000021a82051e20) by thread "Thread-1"
```

Reply:
That's ok - at least we have that covered, and, hopefully it will start to fail from time to time if we regress